File: 2904 - Shortest and Lexicographically Smallest Beautiful String/Method 1/solution.java
class Solution {
    public String shortestBeautifulSubstring(String s, int k) {
        int[] freq = new int[2];
        int n = s.length();
        int low = 0;
        int high = 0;
        int minLen = Integer.MAX_VALUE;
        String str = "";
        while(high<n) {
            int idx = s.charAt(high)-'0';
            freq[idx]++;
            while(freq[1]>k) {
                int lowIdx = s.charAt(low)-'0';
                freq[lowIdx]--;
                low++;
            }
            while(s.charAt(low)=='0' && low<high) {
                freq[0]--;
                low++;
            }
            int len = high - low + 1;
            if(freq[1]==k) {
                if(len<minLen) {
                    minLen = len;
                    str = s.substring(low, high+1);
                } else if(len==minLen) {
                    if(str.length()!=0) {
                        str = str.compareTo(s.substring(low, high+1)) < 0 ? str : s.substring(low, high+1);
                    } else {
                        str = s.substring(low, high+1);
                    }
                }
            }
            high++;
        }
        return str;
    }
}

File: 2904 - Shortest and Lexicographically Smallest Beautiful String/Method 1/solution.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var shortestBeautifulSubstring = function(s, k) {
    let countOne = 0;
    let low = 0;
    let high = 0;
    let n = s.length;
    let str = "";
    let minLen = Number.MAX_SAFE_INTEGER;
    while(high<n) {
        if(s[high]=='1') countOne++;
        while(countOne>k) {
            if(s[low]=='1') countOne--;
            low++;
        }
        while(s[low]=='0' && low<high) {
            low++;
        }
        let len = high - low + 1;
        if(countOne==k) {
            if(len<minLen) {
                minLen = len;
                str = s.substring(low, high+1);
            } else if(len==minLen) {
                if(str.length==0) {
                    str = s.substring(low, high+1);
                } else {
                    str = str < s.substring(low, high+1) ? str : s.substring(low, high+1);
                }
            }
        }
        high++;
    }
    return str;
};